src/adapted/mod.rs

```diff
@@ -257,8 +258,19 @@ impl<'a> CustomTryFrom<raw::client_messages::WriteEventsCompleted<'a>> for Adapt
                     //  * separate (instead of newtype for tuple)
                     // these must be options, as for idempotent writes the positions might not be
                     // returned. both seem to be returned always.
-                    prepare_position: msg.prepare_position.map(|x| x.into()),
-                    commit_position: msg.commit_position.map(|x| x.into()),
+
+                    // FIXME once this sort of error return were removed,
+                    // it will be possible to use the Carrier syntax (?)
```

Comment:
Nope, that's not true, the [`Cow`](https://doc.rust-lang.org/std/borrow/enum.Cow.html) type is here for that, if we did not call [`Cow::to_mut`](https://doc.rust-lang.org/std/borrow/enum.Cow.html#method.to_mut) or [`Cow::into_owned`](https://doc.rust-lang.org/std/borrow/enum.Cow.html#method.into_owned) on it, the data will not be cloned.

The thing to do is to *keep this data borrowed*, we need to be careful with the methods we call to insure we don't use too much ram, because that's ugly.